lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java

```diff
@@ -82,80 +76,66 @@ public Diagnostic validateConfiguration(DiagnosticRequest diagnosticRequest) {
 			// A diagnostic was already added in validatePluginResolution()
 		}
 		if (parameters.isEmpty()) {
-			return null;
+			return;
 		}
 		if (node.isElement() && node.hasChildNodes()) {
 			for (DOMNode childNode : node.getChildren()) {
-				DiagnosticRequest childDiagnosticReq = new DiagnosticRequest(childNode,
-						diagnosticRequest.getXMLDocument(), diagnosticRequest.getDiagnostics());
-				Diagnostic diag = internalValidateConfiguration(childDiagnosticReq, parameters);
-				if (diag != null) {
-					diagnosticRequest.getDiagnostics().add(diag);
-				}
+				DiagnosticRequest childDiagnosticReq = new DiagnosticRequest(childNode, diagnosticRequest);
+				internalValidateConfiguration(childDiagnosticReq, parameters);
 			}
 		}
-		return null;
 	}
 
-	public Diagnostic validateGoal(DiagnosticRequest diagnosticRequest) {
+	public void validateGoal(DiagnosticRequest diagnosticRequest) {
 		DOMNode node = diagnosticRequest.getNode();
 		if (node == null) {
-			return null;
+			return;
 		}
 		if (!validatePluginResolution(diagnosticRequest)) {
-			return null;
+			return;
 		}
 		if (node.isElement() && node.hasChildNodes()) {
 			PluginDescriptor pluginDescriptor;
 			try {
 				pluginDescriptor = MavenPluginUtils.getContainingPluginDescriptor(diagnosticRequest, cache, repoSession,
 						pluginManager);
-				Diagnostic diag = internalValidateGoal(diagnosticRequest, pluginDescriptor);
-				if (diag != null) {
-					diagnosticRequest.getDiagnostics().add(diag);
-				}
+				internalValidateGoal(diagnosticRequest, pluginDescriptor);
+
 			} catch (PluginResolutionException | PluginDescriptorParsingException
 					| InvalidPluginDescriptorException e) {
 				e.printStackTrace();
 				// A diagnostic was already added in validatePluginResolution()
 			}
 		}
-		return null;
 	}
 
-	private static Diagnostic internalValidateGoal(DiagnosticRequest diagnosticReq, PluginDescriptor pluginDescriptor) {
+	private static void internalValidateGoal(DiagnosticRequest diagnosticReq, PluginDescriptor pluginDescriptor) {
```

Comment:
Functions (which don't have side effects) usually leads to easier maintenance than methods like this one that are built on side effects.
So I'd suggest the method becomes something like `private Optional<Diagnostic> validateGoal(DOMNode requestNode, descriptor)`, so consumers could just invoke `validateGoal(node, descriptor).ifPresent(request::addDiagnostic)`. This would also allow easier fine grain testing if necessary.

Reply:
+1 I knew something felt wrong about how the diagnostics were being added.

Will make the fix + other related fixes.